### PR TITLE
[GPU] Einsum with repeated labels and ellipsis support

### PR DESCRIPTION
**Pull request by okhovan**
- Einsum transformation in GPU transformations pipeline with repeated input labels and ellipsis support (just like a reference Einsum implementation);
  - GPU tests added.

### CODE REVIEW

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp [hunk 1]
```diff
@@ -230,60 +204,84 @@ ngraph::Output<ngraph::Node> unsqueeze_input(const ngraph::Output<ngraph::Node>&
...
+    NGRAPH_CHECK(num_broadcasted_dims >= 0);
+
+    LabelDimMap resulted_map;
+    size_t current_dim = 0;
+    for (const auto& label : labels) {
+        if (label == ellipsis) {
+            std::vector<size_t> label_dims;
+            for (size_t ind = 0; ind < num_broadcasted_dims; ++ind) {
+                label_dims.push_back(current_dim + ind);
+            }
```
Comment (tgubanova-lohika) [inline src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp:270]: std::vector<size_t> label_dims(num_broadcasted_dims)
std::iota(l.begin(), l.end(), current_dim);

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp [hunk 2]
```diff
@@ -316,20 +314,20 @@ void transpose_input(ngraph::OutputVector& input_nodes, std::vector<std::string>
 
     // find permutation that establishes bijection between the input subscript
     // and the required one
-    auto labels = ngraph::opset7::Einsum::extract_labels(input_subscript);
-    auto required_labels = ngraph::opset7::Einsum::extract_labels(required_subscript);
+    const auto& input_node = input_nodes[input_ind];
+    const auto labels = ngraph::opset7::Einsum::extract_labels(input_subscript);
+    const auto required_labels = ngraph::opset7::Einsum::extract_labels(required_subscript);
     NGRAPH_CHECK(labels.size() == required_labels.size());
+    auto label_dim_map = compute_label_dim_map(input_node.get_partial_shape().rank(), input_subscript);
```
Comment (tgubanova-lohika) [inline src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp:321]: const?

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp [hunk 3]
```diff
@@ -395,6 +400,211 @@ void reduce_input(ngraph::pass::EinsumDecomposition *einsum_decompose_ptr,
...
+                     const ngraph::Shape& new_common_shape,
+                     const ngraph::Shape& separate_shape,
+                     const ngraph::Shape& reduced_shape,
+                     bool is_separate_first,
+                     ngraph::NodeVector& subgraph_nodes) {
+    NGRAPH_CHECK(input_ind < inputs.size());
+    const auto& input = inputs[input_ind];
+
+    ngraph::Shape new_shape;
+    new_shape.insert(new_shape.end(), new_common_shape.begin(), new_common_shape.end());
```
Comment (tgubanova-lohika) [inline src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp:417]: why not init in counstructor?

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp [hunk 4]
```diff
@@ -395,6 +400,211 @@ void reduce_input(ngraph::pass::EinsumDecomposition *einsum_decompose_ptr,
...
+    }
+    const auto identity = ngraph::opset7::Constant::create(input_node.get_element_type(), identity_shape, identity_values);
+    subgraph_nodes.insert(subgraph_nodes.end(), {identity});
+    return subgraph_nodes.back();
+}
+
+ngraph::Output<ngraph::Node> build_multi_identity(ngraph::pass::EinsumDecomposition* einsum_decompose_ptr,
+                                                  const ngraph::Output<ngraph::Node>& input_node,
+                                                  const std::vector<std::string>& repeated_labels,
+                                                  LabelDimMap& label_dim_map,
```
Comment (tgubanova-lohika) [inline src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp:477]: why not const?

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp [hunk 5]
```diff
@@ -395,6 +400,211 @@ void reduce_input(ngraph::pass::EinsumDecomposition *einsum_decompose_ptr,
...
+    const auto input_shape = input_node.get_shape();
+
+    std::string resultant_subscript = "";
+    static const std::string ellipsis = "...";
+    const auto labels = ngraph::opset7::Einsum::extract_labels(input_subscript);
+    auto label_dim_map = compute_label_dim_map(input_node.get_partial_shape().rank(), input_subscript);
+    std::vector<std::string> repeated_labels;
+    ngraph::Shape result_shape;
+    ngraph::AxisSet reduced_axes;
+    for (const auto& label : labels) {
```
Comment (tgubanova-lohika) [inline src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp:520]: can this loop at least be moved to separate method to make code little bit more readable?

**Review by rkazants — CHANGES_REQUESTED** [review]
Need to validate this PR on Einsum layer tests. Please let me do it on my own today because those tests are not  a part of validation system yet

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp (hunk 4 above)
Comment (rkazants) [inline src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp:477]: I think it is a copy-paste from reference part I have written. But the problem with such approach that the transformation is not reshapable. The transformation makes a model non-reshapable.
We can't accept this implementation

**Comment by rkazants** [comment]
@okhovan, @tgubanova-lohika, guys, let me write you some guidance about how to proceed today. The PR is valuable but it needs some corrections

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp [hunk 6]
```diff
@@ -230,60 +204,84 @@ ngraph::Output<ngraph::Node> unsqueeze_input(const ngraph::Output<ngraph::Node>&
...
-    ngraph::OutputVector new_shape_parts;
-    new_shape_parts.insert(new_shape_parts.end(), common_sub_shape.begin(), common_sub_shape.end());
+ngraph::Output<ngraph::Node> reshape_input_for_matmul(const ngraph::Output<ngraph::Node>& input_node,
+                                                      const ngraph::Shape& common_sub_shape,
+                                                      const ngraph::Shape& separate_sub_shape,
+                                                      const ngraph::Shape& reduced_sub_shape_prod,
+                                                      bool is_separate_first,
+                                                      ngraph::NodeVector& subgraph_nodes) {
+    ngraph::Shape new_shape;
+    new_shape.insert(new_shape.end(), common_sub_shape.begin(), common_sub_shape.end());
```
[the remark this answers is not in the document]
Comment (okhovan) [reply src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp]: Done.

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp [hunk 7]
```diff
@@ -230,60 +204,84 @@ ngraph::Output<ngraph::Node> unsqueeze_input(const ngraph::Output<ngraph::Node>&
...
 
     // compute a product of a sub-shape for separate labels
-    ngraph::OutputVector separate_parts;
-    if (common_sub_shape.size() > 0 && separate_sub_shape.size() == 0) {
+    ngraph::Shape separate_sub_shape_prod = separate_sub_shape;
+    if (common_sub_shape.size() > 0 && separate_sub_shape_prod.empty()) {
```
[the remark this answers is not in the document]
Comment (okhovan) [reply src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp]: Done.

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp (hunk 1 above)
Comment (okhovan) [reply src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp]: Done.

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp (hunk 2 above)
Comment (okhovan) [reply src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp]: Done.

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp (hunk 3 above)
Comment (okhovan) [reply src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp]: Done.

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp (hunk 4 above)
Comment (okhovan) [reply src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp]: Done.

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp (hunk 5 above)
Comment (okhovan) [reply src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp]: Done.

**Comment by okhovan** [comment]
Hi @rkazants 
I'm not sure what exactly _"is not reshapable"_ means. If its about dynamic shapes - AFAIK it should not be blocker, since dynamic shapes is not supported in GPU plugin at the moment.
Looking forward for your comments regarding corrections for this PR.
Thank you!

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp (hunk 4 above)
Comment (tgubanova-lohika) [reply src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp]: We agreed with @p-durandin  to make implementation inside transformations.
If "reshapable" means to provide dynamic shape support - it was not expected in the scope of this PR. Could you clarify a little bit about needed changes?

**Comment by vladimir-paramuzov** [comment]
@tgubanova-lohika  @okhovan "reshapable" means that ov::Model::reshape() method can be called for given model and it can successfully propagate shapes through the whole network. Some operations or some patterns may prevent further shape update. The simplest example is Reshape operation with hardcoded shape parameter which won't be adjusted w.r.t input shape change. So it's not directly connected to dynamic shapes.

@rkazants Do we really need to preserve reshapability in this transformation? When it may be needed?

**Comment by rkazants** [comment]
> @tgubanova-lohika @okhovan "reshapable" means that ov::Model::reshape() method can be called for given model and it can successfully propagate shapes through the whole network. Some operations or some patterns may prevent further shape update. The simplest example is Reshape operation with hardcoded shape parameter which won't be adjusted w.r.t input shape change. So it's not directly connected to dynamic shapes.
> 
> @rkazants Do we really need to preserve reshapability in this transformation? Is it needed for dynamic case only?

Guys, very sorry for the long delay!

This is a common decomposition transformation that is used also by the CPU plugin. We have a requirement to at least common transformation that is to have reshape-able transformation which means transformation executed once during loading a model into a plugin and the resulted sub-graph is used for input data of any shape including dynamic shapes. The current implementation in the master takes this into account.
Now we have a plan to implement in our opset EyeLike operation that actually can resolve a limitation with diagonal case (repeated labels in subscript) in Einsum equation like "aaab,bbbc->ac". After introducing this operation we can decompose this case using EyeLike.

@vladimir-paramuzov, If you do want to have the transformation proposed in this PR, you can create plugin transformation, switch it on and switch off the common transformation.

@GlebKazantaev, @ilyachur, want to check with you guys. Do we still need to have reshape-able transformations in the common part? What is the current concept for API 1.0 and API 2.0?

File: src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp (hunk 4 above)
Comment (rkazants) [reply src/common/transformations/src/transformations/op_conversions/einsum_decomposition.cpp]: @tgubanova-lohika, @p-durandin, sorry but I does not support to break reshapability of this transformations that is used by different plugins. It is not a good idea. You can create a separate transformation for GPU plugin and execute it there.

**Comment by vladimir-paramuzov** [comment]
@rkazants Since we don't have EyeLike operation yet to decompose Einsum in reshapable manner for some cases, does it make sense to merge this PR to enable Einsum decomposition for static shapes, and then make it reshapable once opset is extended?

**Comment by rkazants** [comment]
> @rkazants Since we don't have EyeLike operation yet to decompose Einsum in reshapable manner for some cases, does it make sense to merge this PR to enable Einsum decomposition for static shapes, and then make it reshapable once opset is extended?

We agreed to move this implementation into GPU side. Thanks!

**Comment by tgubanova-lohika** [comment]
> > @rkazants Since we don't have EyeLike operation yet to decompose Einsum in reshapable manner for some cases, does it make sense to merge this PR to enable Einsum decomposition for static shapes, and then make it reshapable once opset is extended?
> 
> We agreed to move this implementation into GPU side. Thanks!

It's already moved to GPU, could you continue the review?

**Review by yeonbok — APPROVED** [review]
LGTM.

**Review by wilson-seok — APPROVED** [review]
Checked the PR with real models - BERT TF2 model, ALBERT model, xlnet model.
Einsum decomposition in GPU plugin worked properly as same as IE common einsum decomposition.

**Comment by wilson-seok** [comment]
@yeonbok Can we remove "do not merge" tag?

**Comment by tgubanova-lohika** [comment]
new PR created https://github.com/openvinotoolkit/openvino/pull/11615

**Pull request closed without merging**